code/game/objects/structures/doors/_door.dm

```diff
@@ -0,0 +1,195 @@
+/obj/structure/door
+	name = "door"
+	icon = 'icons/obj/doors/material_doors.dmi'
+	icon_state = "metal"
+	hitsound = 'sound/weapons/genhit.ogg'
+	material_alteration = MAT_FLAG_ALTERATION_NAME | MAT_FLAG_ALTERATION_COLOR | MAT_FLAG_ALTERATION_COLOR
+	maxhealth = 50
+	density =  TRUE
+	anchored = TRUE
+	opacity =  TRUE
+
+	var/changing_state = FALSE
+	var/icon_base
+	var/datum/lock/lock
+
+/obj/structure/door/Initialize()
+	. = ..()
+	if(!istype(material))
+		return INITIALIZE_HINT_QDEL
+	if(lock)
+		lock = new /datum/lock(src, lock)
+	if(!icon_base)
+		icon_base = material.door_icon_base
+	update_nearby_tiles(need_rebuild=TRUE)
+
+/obj/structure/door/Destroy()
+	update_nearby_tiles()
+	QDEL_NULL(lock)
+	return ..()
+
+/obj/structure/door/get_material_health_modifier()
+	. = 10
+
+/obj/structure/door/on_update_icon()
+	..()
+	if(density)
+		icon_state = "[icon_base]"
+	else
+		icon_state = "[icon_base]open"
+
+/obj/structure/door/proc/post_change_state()
+	update_nearby_tiles()
+	update_icon()
+	changing_state = FALSE
+
+/obj/structure/door/attack_hand(var/mob/user)
+	return density ? open() : close()
+
+/obj/structure/door/proc/close()
+	set waitfor = 0
+	if(!can_close())
+		return FALSE
+	flick("[icon_base]closing", src)
+	playsound(src.loc, material.dooropen_noise, 100, 1)
+
+	changing_state = TRUE
+	sleep(1 SECOND)
+	set_density(TRUE)
+	set_opacity(initial(opacity) && material.opacity > 0.5)
```

Comment:
If I understand the logic correctly, this means that I can make steel doors which are transparent when closed, but not glass doors which are opaque when closed? Seems kind of odd.

Reply:
I suppose I can add a flag for non-opaque doors. Nothing will actually use it though.